metals/src/main/scala/scala/meta/internal/parsing/JavaTokens.scala

```diff
@@ -0,0 +1,89 @@
+package scala.meta.internal.parsing
+
+import scala.collection.compat.immutable.ArraySeq
+import scala.collection.mutable.ArrayBuffer
+import scala.util.Failure
+import scala.util.Success
+import scala.util.Try
+
+import scala.meta.inputs.Input
+import scala.meta.inputs.Input.File
+import scala.meta.inputs.Input.VirtualFile
+
+import org.eclipse.jdt.core.ToolFactory
+import org.eclipse.jdt.core.compiler.IScanner
+import org.eclipse.jdt.core.compiler.ITerminalSymbols
+
+object JavaTokens {
+
+  /**
+   * Best effort translation of java tokens to the scalameta tokens, should work for most purposes.
+   *
+   * @param source to tokenize
+   * @return tokenized source
+   */
+  def tokenize(source: Input): Option[Seq[JavaToken]] = Try {
+
+    val scanner = ToolFactory.createScanner(
+      /*tokenizeComments = */ true, /*tokenizeWhiteSpace = */ true,
+      /*assertMode = */ false, /*recordLineSeparator = */ true
+    )
+    scanner.setSource(source.chars)
+
+    val buffer = new ArrayBuffer[JavaToken]()
+
+    def loop(token: Int): Unit = {
+      if (token != ITerminalSymbols.TokenNameEOF) {
+        mapToken(token, scanner, source, buffer)
+        loop(scanner.getNextToken())
+      }
+    }
+
+    loop(scanner.getNextToken())
+    ArraySeq(buffer: _*)
+  } match {
+    case Failure(exception) =>
+      source match {
+        case VirtualFile(path, _) =>
+          scribe.warn(s"Cannot tokenize Java file $path", exception)
+        case File(path, _) =>
+          scribe.warn(s"Cannot tokenize Java file $path", exception)
+        case _ =>
+      }
+      None
+    case Success(value) => Some(value)
+  }
+
+  private def mapToken(
+      i: Int,
+      scanner: IScanner,
+      source: Input,
+      buffer: ArrayBuffer[JavaToken]
+  ): Unit = {
```

Comment:
Could `mapToken` return a token instead of adding it to the buffer? It's always better to have fewer side effects.

Reply:
I wanted to make it more effective, since this is invoked a lot of the times.

Reply:
Inlined this anyway so no more side effects from that function!